contracts/inverse-bonds/types/Pro/Viewer.sol

```diff
@@ -0,0 +1,245 @@
+   * @notice             should be used if length exceeds max to query entire array
+   */
+    function liveMarketsBetween(uint256 firstIndex, uint256 lastIndex) public view returns (uint256[] memory) {
+        uint256 num;
+        for (uint256 i = firstIndex; i < lastIndex; i++) {
+            if (isLive(i)) num++;
+        }
+
+        uint256[] memory ids = new uint256[](num);
+        uint256 nonce;
+        for (uint256 i = firstIndex; i < lastIndex; i++) {
+            if (isLive(i)) {
+                ids[nonce] = i;
+                nonce++;
+            }
+        }
+        return ids;
+    }
+
+    /**
+     * @notice             returns an array of all active market IDs for a given quote token
+   * @param _creator     is query for markets by creator, or for markets by token
+   * @param _base        if query is for markets by token, search by base or quote token
+   * @param _address     address of creator or token to query by
+   */
+    function liveMarketsFor(bool _creator, bool _base, address _address) public view override returns (uint256[] memory) {
+        uint256[] memory mkts;
+
+        if (_creator) {
+            mkts = marketsForCreator[_address];
+        } else {
+            mkts = _base
+            ? marketsForBase[_address]
+            : marketsForQuote[_address];
+        }
+
+        uint256 num;
+
+        for (uint256 i = 0; i < mkts.length; i++) {
+            if (isLive(mkts[i])) num++;
+        }
+
+        uint256[] memory ids = new uint256[](num);
+        uint256 nonce;
+
+        for (uint256 i = 0; i < mkts.length; i++) {
+            if (isLive(mkts[i])) {
+                ids[nonce] = mkts[i];
+                nonce++;
+            }
+        }
+        return ids;
+    }
+
+
+    function marketsFor(address tokenIn, address tokenOut) public view returns (uint256[] memory) {
+        uint256[] memory forBase = liveMarketsFor(false, true, tokenOut);
+        uint256[] memory ids;
+        uint256 nonce;
+        for(uint256 i; i < forBase.length; i++) {
+            if (address(markets[forBase[i]].quoteToken) == tokenIn) {
+                ids[nonce] = forBase[i];
+            }
+        }
+        return ids;
+    }
+
+    function findMarketFor(
+        address tokenIn,
+        address tokenOut,
+        uint256 amountIn,
+        uint256 minAmountOut,
+        uint256 maxExpiry
+    ) external view returns (uint256 id) {
+        uint256[] memory ids = marketsFor(tokenIn, tokenOut);
+        uint256[] memory payouts;
+        uint256 n;
+        for(uint256 i; i < ids.length; i++) {
+            Terms memory term = terms[ids[i]];
+
+            uint256 expiry = term.fixedTerm ? block.timestamp + term.vesting : term.vesting;
+            require(expiry <= maxExpiry, "Bad expiry");
+
+            if (minAmountOut > markets[ids[i]].maxPayout) {
+                payouts[n] = payoutFor(amountIn, ids[i]);
+            } else {
+                payouts[n] = 0;
+            }
```

Comment:
This might need more eyes on this but I believe this whole `if/else` is wrong. Lets assume this inverse bond is `KLIMA/USDC` thus we put in `KLIMA` and we get `USDC`. The `minAmountOut` here is in `USDC` as well as the `maxPayout` and so lets take an example.

If `minAmountOut = 100` and `maxPayout = 50`, then the we set the `payouts[n]` to the payout but `payoutFor()` function doesn't cap the value onto capacity of 

If the`minAmountOut` is **above** the `maxPayout`, then it seems to me that it should be set to `0` as in that market won't be able to satisfy the minimum required.

If the`minAmountOut` is **below** the `maxPayout`, then that should be `payoutFor()` used.

In other words, this seems reverse?